libc/string_cpy.c

```diff
@@ -0,0 +1,1269 @@
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(input, testArray, MAX_ASCII_SET);
+}
+
+
+TEST(string_memcpy, nothing)
+{
+	char testArray[MAX_STR_LEN] = TEST_STR1,
+		 input[MAX_STR_LEN] = { 0 },
+		 strText[] = "Lorem ipsum \0 dolor";
+
+	/* Clearing array with support array filled with 0 */
+	TEST_ASSERT_EQUAL_PTR(testArray, memcpy(testArray, input, MAX_STR_LEN));
+	TEST_ASSERT_NOT_EQUAL_INT(0, strcmp(testArray, TEST_STR1));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(input, testArray, MAX_STR_LEN);
+
+
+	/* Checking no sensitivity for NULL */
+	memset(input, 0, sizeof(input));
+	memcpy(input, strText, sizeof(strText));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(strText, input, sizeof(strText));
+}
+
+
+TEST(string_memcpy, big)
+{
+	char buff[1024] = { 0 };
+
+	/* Checking capability of handling big blocks of data */
+	TEST_ASSERT_EQUAL_INT(1, big_string());
+	TEST_ASSERT_EQUAL_PTR(&buff, memcpy(buff, long_str, 1024));
+	TEST_ASSERT_EQUAL_STRING(buff, long_str);
+	TEST_ASSERT_EQUAL_CHAR(buff[0], '\1');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 4], 'A');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 2], '\x81');
+
+	/* Offset big chunk of data */
+	TEST_ASSERT_EQUAL_PTR(&buff[128], memcpy(&buff[128], long_str, 1024 - 128));
+	TEST_ASSERT_EQUAL_CHAR(buff[0], '\1');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 4], 'A');
+	TEST_ASSERT_EQUAL_CHAR(buff[MAX_ASCII_SET / 2], '\x01');
+	TEST_ASSERT_NOT_EQUAL_CHAR(buff[MAX_ASCII_SET / 2], '\x7F');
+	TEST_ASSERT_NOT_EQUAL_CHAR(buff[MAX_ASCII_SET], '\x01');
+	TEST_ASSERT_EQUAL_CHAR(buff[(MAX_ASCII_SET / 4) + 128], 'A');
+	TEST_ASSERT_EQUAL_CHAR(buff[(MAX_ASCII_SET / 2) + 128], '\x81');
+}
+
+
+/*
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+ */
+
+
+TEST_SETUP(string_memccpy)
+{
+}
+
+
+TEST_TEAR_DOWN(string_memccpy)
+{
+}
+
+
+TEST(string_memccpy, basic)
+{
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+#ifndef __phoenix__
+
+	char strDest[MAX_STR_LEN] = { 0 },
+		 chr = 'c',
+		 chrDest,
+		 strPlaceholder[MAX_STR_LEN] = TEST_STR2;
+
+
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(strDest, TEST_STR1, 'x', sizeof(TEST_STR1)));
+	TEST_ASSERT_EQUAL_STRING(TEST_STR1, strDest);
+
+
+	memset(strDest, 0, sizeof(strDest));
+	sprintf(strDest, "%s", TEST_STR1);
+
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(strPlaceholder, strDest, 'x', sizeof(strDest)));
+	TEST_ASSERT_EQUAL_STRING(TEST_STR1, strDest);
+	TEST_ASSERT_EQUAL_STRING(strPlaceholder, TEST_STR1);
+
+	/* Checking if strDest doesn't hold old data from str */
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(strDest, TEST_STR3, 'x', sizeof(TEST_STR3)));
+	TEST_ASSERT_EQUAL_STRING(TEST_STR3, strDest);
+	TEST_ASSERT_NOT_EQUAL_CHAR(strDest[0], TEST_STR1[0]);
+	TEST_ASSERT_NOT_EQUAL_INT(0, strcmp(strDest, TEST_STR1));
+
+	/* Checking ability to copy one char */
+	TEST_ASSERT_EQUAL_PTR(NULL, memccpy(&chrDest, &chr, 'x', sizeof(chr)));
+	TEST_ASSERT_NOT_EMPTY(&chrDest);
+	TEST_ASSERT_EQUAL_CHAR(chr, chrDest);
+
```

Comment:
You could also create a simple check, which will copy less bytes than available in the buffer.